x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java

```diff
@@ -404,4 +444,76 @@ void doExecute(Action<Request, Response, RequestBuilder> action, Request request
         return storedContext;
     }
 
+    static final class IndexMetadataVersionChecker implements BiConsumer<Long, Consumer<Exception>> {
+
+        private static final Logger LOGGER = Loggers.getLogger(IndexMetadataVersionChecker.class);
+
+        private final Client followClient;
+        private final Client leaderClient;
+        private final Index leaderIndex;
+        private final Index followIndex;
+        private final AtomicLong currentIndexMetadataVersion;
+        private final Semaphore updateMappingSemaphore = new Semaphore(1);
+
+        IndexMetadataVersionChecker(Index leaderIndex, Index followIndex, Client followClient, Client leaderClient) {
+            this.followClient = followClient;
+            this.leaderIndex = leaderIndex;
+            this.followIndex = followIndex;
+            this.leaderClient = leaderClient;
+            this.currentIndexMetadataVersion = new AtomicLong();
+        }
+
+        public void accept(Long minimumRequiredIndexMetadataVersion, Consumer<Exception> handler) {
+            if (currentIndexMetadataVersion.get() >= minimumRequiredIndexMetadataVersion) {
+                LOGGER.debug("Current index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                    currentIndexMetadataVersion.get(), minimumRequiredIndexMetadataVersion);
+                handler.accept(null);
+            } else {
+                updateMapping(minimumRequiredIndexMetadataVersion, handler);
+            }
+        }
+
+        void updateMapping(long minimumRequiredIndexMetadataVersion, Consumer<Exception> handler) {
+            try {
+                updateMappingSemaphore.acquire();
+            } catch (InterruptedException e) {
+                handler.accept(e);
+                return;
+            }
+            if (currentIndexMetadataVersion.get() >= minimumRequiredIndexMetadataVersion) {
+                updateMappingSemaphore.release();
+                LOGGER.debug("Current index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                    currentIndexMetadataVersion.get(), minimumRequiredIndexMetadataVersion);
+                handler.accept(null);
+                return;
+            }
+
+            ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+            clusterStateRequest.clear();
+            clusterStateRequest.metaData(true);
+            clusterStateRequest.indices(leaderIndex.getName());
+
+            leaderClient.admin().cluster().state(clusterStateRequest, ActionListener.wrap(clusterStateResponse -> {
+                IndexMetaData indexMetaData = clusterStateResponse.getState().metaData().getIndexSafe(leaderIndex);
```

Comment:
as follow up somewhere we need to make a note to only fetch the right index metadata - this can lead to transferring 500MB of data  (the entire cluster state metadata ) for one index.

Reply:
@bleskes Only the IMD of a single index is being fetched now: https://github.com/elastic/elasticsearch/pull/30115/files#diff-27bdc34d556e302e5a2f3801100e1af7R494

Reply:
@martijnvg awesome. thx. Also we should probably validate the index uuid.

Reply:
Sorry, I was too terse. I was saying this in the context of https://github.com/elastic/elasticsearch/pull/30115/files#diff-27bdc34d556e302e5a2f3801100e1af7R379